clickhouse-admin/src/context.rs

```diff
@@ -3,84 +3,555 @@
+                clickward,
+                log,
+                node_settings,
+                response,
+            } => {
+                let result = generate_config_and_enable_svc(
+                    generation_tx.clone(),
+                    clickward,
+                    node_settings,
+                );
+                if let Err(e) = response.send(result) {
+                    error!(
+                        &log,
+                        "failed to send value from configuration generation to channel: {e:?}"
+                    );
+                };
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum NodeSettings {
+    Replica { settings: ServerConfigurableSettings, fmri: String },
+    Keeper { settings: KeeperConfigurableSettings, fmri: String },
+}
+
+impl NodeSettings {
+    fn fmri(self) -> String {
+        match self {
+            NodeSettings::Replica { fmri, .. } => fmri,
+            NodeSettings::Keeper { fmri, .. } => fmri,
+        }
+    }
+
+    fn generation(&self) -> Generation {
+        match self {
+            NodeSettings::Replica { settings, .. } => settings.generation(),
+            NodeSettings::Keeper { settings, .. } => settings.generation(),
+        }
+    }
+}
+
+pub fn generate_config_and_enable_svc(
+    generation_tx: watch::Sender<Option<Generation>>,
+    clickward: Clickward,
+    node_settings: NodeSettings,
+) -> Result<GenerateConfigResult, HttpError> {
+    let incoming_generation = node_settings.generation();
+    let generation_rx = generation_tx.subscribe();
+    let current_generation = *generation_rx.borrow();
+
+    // If the incoming generation number is lower, then we have a problem.
+    // We should return an error instead of silently skipping the configuration
+    // file generation.
+    if let Some(current) = current_generation {
+        if current > incoming_generation {
+            return Err(HttpError::for_client_error(
+                Some(String::from("Conflict")),
+                StatusCode::CONFLICT,
+                format!(
+                    "current generation '{}' is greater than incoming generation '{}'",
+                    current,
+                    incoming_generation,
+                )
+            ));
+        }
+    };
+
+    let output = match node_settings {
+        NodeSettings::Replica { ref settings, .. } => {
+            GenerateConfigResult::Replica(
+                clickward.generate_server_config(settings)?,
+            )
+        }
+        NodeSettings::Keeper { ref settings, .. } => {
+            GenerateConfigResult::Keeper(
+                clickward.generate_keeper_config(settings)?,
+            )
+        }
+    };
+
+    // We want to update the generation number only if the config file has been
+    // generated successfully.
+    generation_tx.send(Some(incoming_generation)).map_err(|e| {
+        HttpError::for_internal_error(format!("failure to send request: {e}"))
+    })?;
```

Comment:
This is fine but kind of awkward: `send` can only fail if there are no subscribers, which in our case would mean the context object is gone, which presumably means there isn't anyone around to receive the HTTP error we're creating.

`watch::Sender` has a few methods that let you update the value without failing even if there are no receivers (`send_replace`, `send_if_modified`, `send_modify`). Since `Generation` is basically just an integer, `send_replace` might be the cleanest here?

```suggestion
    generation_tx.send_replace(Some(incoming_generation));
```